crypto/threads_common.c

```diff
@@ -388,8 +402,39 @@ int CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_KEY_ID id,
      *
      * Assign to the entry in the table so that we can find it later
      */
-    return ossl_sa_CTX_TABLE_ENTRY_set(mkey[id].ctx_table,
-                                       (uintptr_t)ctx, data);
+    ret = ossl_sa_CTX_TABLE_ENTRY_set(mkey->entries[id].ctx_table,
+                                      (uintptr_t)ctx, data);
+
+    /*
+     * This is done so that on library cleanup, we ensure that thread local storage for our
+     * master key is deallocated - i.e. like other thread local storage that uses system level keys,
+     * we want to have zero allocated heap after OPENSSL_cleanup is called.
+     * Note we only do this for the non-fips build case, as FIPS Thread local storage is reaped via
+     * CRYPTO_THREAD_clean_for_fips, so we don't need to worry about it here.
+     * But we need to be very careful about a few things:
+     * 1) We need to only call this at the end of this function, and only if we allocated a new
+     *    mkey.  The latter constraint is because we don't want to register a new thread cleanup
+     *    routine when we allocate the mkey, to avoid multiple cleanup handler.  The latter
+     *    is because calling this function may recurse back to here, as this function calls
+     *    CRYPTO_THREAD_set_local_ex(TEVENT_KEY).  Calling this right before we return ensures
+     *    that any such recursion will find the key that we set here, and not loop forever.
+     *
+     * 2) We need to call this at the maximum loweset priority (i.e. this cleanup handler MUST
```

Comment:
this makes me think if we should disallow for users to set UINT_MAX and set a limit of `UINT_MAX -1` (or even keep more reserved numbers if you can think of any use in the future of it) to avoid UB when the user sets UINT_MAX priority and the handlers would be called in an unspecified order causing possibility that `thread_skip_clear_mkey` wouldn't be the last one

Reply:
thats a reasonable argument.